teps/0145-cel-in-whenexpression.md

```diff
@@ -0,0 +1,372 @@
+  - [Related Work](#related-work)
+- [Proposal](#proposal)
+  - [Syntax](#syntax)
+  - [Validation](#validation)
+  - [Example](#example)
+- [Design Evaluation](#design-evaluation)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [Conformance](#conformance)
+  - [User Experience](#user-experience)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes to add CEL (Common Expression Language) support in `when` expressions to enable more powerful conditional execution and improve the user experience by offering a more concise way of defining conditionals.
+
+
+## Motivation
+
+CEL is a non-Turing complete language designed for simplicity, speed, safety, and portability. Tekton has adopted CEL expressions in [Trigger] to match specific events to trigger PipelineRuns, and [CELCustomRun] is implemented to experiment with CEL in Tekton Pipelines without adding CEL directly to the Tekton API.
+
+There are several feature requests in the community which can be met if CEL is supported in Tekton Pipeline:
+- [#3591] calls for supporting ANY match behaviour in `WhenExpressions` to execute the task if any `params` is not empty.
+- [#3149] suggests using CEL to split the string instead of adding an additional task to do the work.
+
+Take a look at current [WhenExpressions]:
+
+```yaml
+when:
+  - input: "$(params.path)"
+    operator: in
+    values: ["README.md"]
+```
+
+Current `WhenExpression` requires users to define three fields: `input`, `operator` and `values`. but CEL expressions can achieve the same functionality with a single string. Additionally, the current operator only supports `in` and `notin`, which can be limiting for complex use cases. For example, users will need to figure out [hacky workarounds][#3591] to support any in their case.
+
+For these reasons, `WhenExpressions` would be a good place to start introducing CEL in Tekton Pipelines. This would allow users to write more concise and powerful conditional expressions, and it could be extended to other fields in the future.
+
+### Use Cases
+
+- Cover all the functionalities of current `WhenExpression` and offer a simpler way of having the conditional execution
+- Support ANY: e.g. Execute the task with if any params is not empty
+- Support numeric operators such as GreaterThan/LessThan, e.g. Execute the task if test coverage is large than a certain number.
+
+### Goals
+- Support CEL in `WhenExpression`.
+
+### Non-Goals
+- Support CEL in Pipeline API fields other than `WhenExpression`.
+
+### Requirements
+- The syntax should be extensible to support other expression languages in the future if needed.
+- Users cannot inject CEL via variables such as params and results.
```

Comment:
Is this temporary or will be restricted in future too? 

I do not have any specific use case but trying to evaluate and understand if we are open to this in future on a need basis.

Reply:
Did you mean the `Users cannot inject CEL via variables such as params and results.` We think this will be restricted in the future as well for security concerns. The problem is similar to https://github.com/tektoncd/community/pull/1077

Reply:
This is a little ambiguous, suggest rephrasing this requirement to:

```
The pipeline authors cannot inject an entire CEL expression using a variable such as params or results i.e. CEL expression can include params and results which will be resolved but params and results cannot hold an entire valid CEL expression. This requirement is imposed due to a security concern as described in [TEP-0146](https://github.com/tektoncd/community/pull/1077).
```